From fbe5e855d2b3bc859653be445883f8c73b41b230 Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Thu, 16 Jan 2025 14:09:40 +0100
Subject: [PATCH] tapdisk: protect td_blktap_t structure

libqcow2 spawns a thread to handle IO requests to qcow2 layer.
Add a mutex to protect blktap structure from qcow2 concurrency.

This is the second part of the tapdisk threading protection.

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 drivers/tapdisk-blktap.c | 34 ++++++++++++++++++++++++----------
 drivers/tapdisk-blktap.h |  1 +
 2 files changed, 25 insertions(+), 10 deletions(-)

diff --git a/drivers/tapdisk-blktap.c b/drivers/tapdisk-blktap.c
index cf72b095..9c2e4855 100644
--- a/drivers/tapdisk-blktap.c
+++ b/drivers/tapdisk-blktap.c
@@ -87,19 +87,21 @@ struct td_blktap_req {
         struct timeval          ts;
 };
 
-td_blktap_req_t *
+static td_blktap_req_t *
 tapdisk_blktap_alloc_request(td_blktap_t *tap)
 {
 	td_blktap_req_t *req = NULL;
 
+	pthread_mutex_lock(&tap->mutex);
 	if (likely(tap->n_reqs_free))
 		req = tap->reqs_free[--tap->n_reqs_free];
+	pthread_mutex_unlock(&tap->mutex);
 
 	return req;
 }
 
-void
-tapdisk_blktap_free_request(td_blktap_t *tap, td_blktap_req_t *req)
+static void
+tapdisk_blktap_free_request_locked(td_blktap_t *tap, td_blktap_req_t *req)
 {
 	BUG_ON(tap->n_reqs_free >= tap->n_reqs);
 	tap->reqs_free[tap->n_reqs_free++] = req;
@@ -117,6 +119,8 @@ tapdisk_blktap_reqs_free(td_blktap_t *tap)
 		free(tap->reqs_free);
 		tap->reqs_free = NULL;
 	}
+
+	pthread_mutex_destroy(&tap->mutex);
 }
 
 static int
@@ -124,6 +128,8 @@ tapdisk_blktap_reqs_init(td_blktap_t *tap, int n_reqs)
 {
 	int i, err;
 
+	pthread_mutex_init(&tap->mutex, NULL);
+
 	tap->reqs = malloc(n_reqs * sizeof(td_blktap_req_t));
 	if (!tap->reqs) {
 		err = -errno;
@@ -139,8 +145,10 @@ tapdisk_blktap_reqs_init(td_blktap_t *tap, int n_reqs)
 	tap->n_reqs      = n_reqs;
 	tap->n_reqs_free = 0;
 
+	pthread_mutex_lock(&tap->mutex);
 	for (i = 0; i < n_reqs; i++)
-		tapdisk_blktap_free_request(tap, &tap->reqs[i]);
+		tapdisk_blktap_free_request_locked(tap, &tap->reqs[i]);
+	pthread_mutex_unlock(&tap->mutex);
 
 	return 0;
 
@@ -180,7 +188,7 @@ tapdisk_blktap_error_status(td_blktap_t *tap, int error)
 }
 
 static void
-__tapdisk_blktap_push_response(td_blktap_t *tap, int final)
+__tapdisk_blktap_push_response_locked(td_blktap_t *tap, int final)
 {
 	tap->rsp_prod_pvt++;
 
@@ -200,13 +208,15 @@ tapdisk_blktap_fail_request(td_blktap_t *tap,
 
 	BUG_ON(!tap->vma);
 
+	pthread_mutex_lock(&tap->mutex);
 	rsp = BLKTAP_GET_RESPONSE(tap, tap->rsp_prod_pvt);
 
 	rsp->id        = msg->id;
 	rsp->operation = msg->operation;
 	rsp->status    = tapdisk_blktap_error_status(tap, error);
 
-	__tapdisk_blktap_push_response(tap, 1);
+	__tapdisk_blktap_push_response_locked(tap, 1);
+	pthread_mutex_unlock(&tap->mutex);
 }
 
 static void
@@ -218,6 +228,7 @@ tapdisk_blktap_put_response(td_blktap_t *tap,
         unsigned long long interval;
         struct timeval now;
 
+	pthread_mutex_lock(&tap->mutex);
 	BUG_ON(!tap->vma);
 
 	rsp = BLKTAP_GET_RESPONSE(tap, tap->rsp_prod_pvt);
@@ -246,7 +257,10 @@ tapdisk_blktap_put_response(td_blktap_t *tap,
 	rsp->operation = op;
 	rsp->status    = tapdisk_blktap_error_status(tap, error);
 
-	__tapdisk_blktap_push_response(tap, final);
+	tapdisk_blktap_free_request_locked(tap, req);
+
+	__tapdisk_blktap_push_response_locked(tap, final);
+	pthread_mutex_unlock(&tap->mutex);
 }
 
 static void
@@ -256,8 +270,6 @@ tapdisk_blktap_complete_request(td_blktap_t *tap,
 {
 	if (likely(tap->vma))
 		tapdisk_blktap_put_response(tap, req, error, final);
-
-	tapdisk_blktap_free_request(tap, req);
 }
 
 static void
@@ -390,7 +402,9 @@ tapdisk_blktap_get_requests(td_blktap_t *tap)
 		err = tapdisk_blktap_parse_request(tap, msg, req);
 		if (err) {
 			tapdisk_blktap_fail_request(tap, msg, err);
-			tapdisk_blktap_free_request(tap, req);
+			pthread_mutex_lock(&tap->mutex);
+			tapdisk_blktap_free_request_locked(tap, req);
+			pthread_mutex_unlock(&tap->mutex);
 			goto fail_ring;
 		}
 
diff --git a/drivers/tapdisk-blktap.h b/drivers/tapdisk-blktap.h
index ca576a77..5bd4aae8 100644
--- a/drivers/tapdisk-blktap.h
+++ b/drivers/tapdisk-blktap.h
@@ -70,6 +70,7 @@ struct td_blktap {
 	td_blktap_req_t        *reqs;
 	int                     n_reqs_free;
 	td_blktap_req_t       **reqs_free;
+	pthread_mutex_t         mutex;
 
 	struct list_head        entry;
 
